test/rules/implicit_reopen_test.dart

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:test_reflective_loader/test_reflective_loader.dart';
+
+import '../rule_test_support.dart';
+
+main() {
+  defineReflectiveSuite(() {
+    defineReflectiveTests(ImplicitReopenTest);
+  });
+}
+
+@reflectiveTest
+class ImplicitReopenTest extends LintRuleTest {
+  @override
+  bool get addMetaPackageDep => true;
+
+  @override
+  List<String> get experiments => ['class-modifiers', 'sealed-class'];
+
+  @override
+  String get lintRule => 'implicit_reopen';
+
+  test_extends_class_classFinal_ok() async {
+    await assertNoDiagnostics(r'''
+final class F {}
+
+class C extends F {}
+''');
+  }
+
+  test_extends_class_classInterface() async {
+    await assertDiagnostics(r'''
+interface class I {}
+
+class C extends I {}
+''', [
+      lint(28, 1),
+    ]);
+  }
+
+  test_extends_class_classInterface_outsideLib_ok() async {
+    newFile('$testPackageLibPath/a.dart', r'''
+interface class I {}
+''');
+
+    await assertDiagnostics(r'''
+import 'a.dart';
+
+class C extends I {}
+''', [
+      error(CompileTimeErrorCode.INTERFACE_CLASS_EXTENDED_OUTSIDE_OF_LIBRARY,
+          34, 1),
+    ]);
+  }
+
+  test_extends_class_classInterface_reopened_ok() async {
+    await assertNoDiagnostics(r'''
+import 'package:meta/meta.dart';
+
+interface class I {}
+
+@reopen
+class C extends I {}
+''');
+  }
+
+  test_extends_class_classSealed_classInterface() async {
+    await assertDiagnostics(r'''
+interface class I {}
+
+sealed class S extends I {}
+
+class C extends S {}
+''', [
+      lint(57, 1),
+    ]);
+  }
+
+  test_extends_class_classSealed_mixinInterface() async {
+    await assertDiagnostics(r'''
+interface class I {}
+
+sealed class S extends I {}
+
+class C extends S {}
+''', [
+      lint(57, 1),
+    ]);
+  }
+
+  test_extends_classBase_classFinal() async {
+    await assertDiagnostics(r'''
+final class F {}
+
+base class B extends F {}
+''', [
+      lint(29, 1),
+    ]);
+  }
+
+  test_extends_classBase_classInterface() async {
+    await assertDiagnostics(r'''
+interface class I {}
+
+base class B extends I {}
+''', [
+      lint(33, 1),
+    ]);
+  }
+
+  test_extends_classBase_classSealed_classFinal() async {
+    await assertDiagnostics(r'''
+final class F {}
```

Comment:
I think you may be missing tests for the second half of the `isInducedFinal` check which is `it has a direct superinterface which is 'base' as well as a direct 'extends' or 'with' superinterface which is 'interface'.`

Reply:
Took another crack at tests but missing cases would be most welcome!